QueryChat + ggsql blog post (#244)
posit-dev/open-source-website: content/blog/querychat-ggsql/index.qmd

---
title: "Ask more of your dashboard with querychat and ggsql"
date: 2026-06-17
people:
  - Carson Sievert
description: >
  querychat now supports ggsql-powered visualizations — users can ask "show me" questions and get charts rendered directly in the chat, alongside SQL queries and reactive filtering, all without arbitrary code execution.
image: featured.png
image-alt: "A querychat app showing a natural language conversation with inline ggsql visualizations and SQL query results"
topics:
  - Artificial Intelligence
  - Interactive Apps
  - Visualization
software:
  - querychat
languages:
  - R
  - Python
nohero: false
hidesubscription: false
---

> I'm super excited to announce that [querychat](https://posit-dev.github.io/querychat/) now supports [ggsql](/blog/2026-04-20_ggsql_alpha_release/). As a result, querychat users can now go beyond predetermined dashboard views and ask the LLM for bespoke visualizations on the fly. With this addition, `querychat` now comes with three pre-built tools that LLMs can use in response to user questions:

Reviewer: ```suggestion
I'm super excited to announce that [querychat](https://posit-dev.github.io/querychat/) now supports [ggsql](/blog/2026-04-20_ggsql_alpha_release/). As a result, querychat users can now go beyond predetermined dashboard views and ask the LLM for bespoke visualizations on the fly. With this addition, querychat now comes with three pre-built tools that LLMs can use in response to user questions:
```

Probably want to have an LLM ensure that you consistently refer to package names without backticks.

1. **Visualize:** execute `ggsql` queries, rendered as visualizations in the chat.
2. **Query:** execute SQL queries, rendered as tables/text in the chat.
3. **Filter:** execute filter SQL queries on predetermined views, allowing the LLM to reactively drill into the data in response to user questions.

> Equipped with these tools, your dashboard can offer a balance between "curated insights" and "explorable data" — surfacing interesting trends and summaries upfront, but also letting users ask follow-up questions that go beyond the predetermined views. The video below gives you a feel for this workflow in action.[^1] Note how the conversation starts off by filtering the predetermined views, but then progresses into more general exploration. 

Reviewer: ```suggestion
Equipped with these tools, your dashboard can offer a balance between "curated insights" and "explorable data" — surfacing interesting trends and summaries upfront, but also letting users ask follow-up questions that go beyond the predetermined views. The video below gives you a feel for this workflow in action[^1]. Note how the conversation starts off by filtering the predetermined views, but then progresses into more general exploration. 
```

[^1]: [Source code](https://gist.github.com/cpsievert/ca0d7e4637fdf994671c9d9fc90cd89f) is available for reference, though you'll need your own Snowflake account and credentials to run it.

{{< video src="querychat-ggsql.mp4" title="querychat in action — querying, filtering, and visualizing data through natural language." >}}

Crucially, every one of `querychat`'s tools works by executing SQL (or `ggsql`) — never arbitrary code. Every query is also fully visible and reproducible: users can inspect the code behind any response, and copy it to run independently. In the next version of `querychat`, you can also expect to see a rich export capability, allowing you to save the insights you care about in a format you care about (e.g. a Quarto dashboard, Jupyter notebook, or R Markdown file).

> Restricting `querychat`'s code execution capability to SQL (or `ggsql`) is an intentional design choice. SQL can be locked down so that the LLM literally cannot modify your data, making it viable for production environments where security and control aren't optional. As a result, `querychat` may not be as capable as a general coding assistant like [Posit Assistant](https://posit-dev.github.io/assistant), but you generally don't want production apps that let users execute arbitrary code anyway. By focusing on SQL execution, `querychat` can maintain a strong security posture while still enabling useful exploration of data via natural language. 

Reviewer: ```suggestion
Restricting `querychat`'s code execution capability to SQL (or `ggsql`) is an intentional design choice. We can guarantee that querychat does not run arbitrary R or python code, making it viable for production environments where security and control are required. As a result, `querychat` may not be as capable as a general coding assistant like [Posit Assistant](https://posit-dev.github.io/assistant), but you generally don't want production apps that let users execute arbitrary code. By focusing on SQL execution, `querychat` can maintain a strong security posture while still enabling useful exploration of data via natural language. 
```

    card(card_header("Measurements"), DTOutput("table"))
  )
)

server <- function(input, output, session) {
  qc_vals <- qc$server()
  df <- qc_vals$df
  output$count <- renderText(nrow(df()))
  output$species <- renderText(length(unique(df()$species)))
  output$avg_mass <- renderText(paste0(round(mean(df()$body_mass_g, na.rm = TRUE)), "g"))
  output$scatter <- renderPlot({
    ggplot(df(), aes(bill_length_mm, bill_depth_mm, color = species)) +
      geom_point()
  })
  output$table <- renderDT(df())
}

shinyApp(ui, server)
```

:::

<img src="dashboard.png" alt="QueryChat embedded in a dashboard with value boxes, a scatter plot, and a data table — all reactively driven by the chat's filter state" class="shadow rounded" />

::: callout-note
### Python users: Shiny v Streamlit/Dash/Gradio

`querychat` also supports `streamlit`, `dash`, and `gradio`. That said, the reactive `df()` pattern shown above — where the LLM's filter state automatically drives every view in the app — is what Shiny's reactive programming model was designed for. In other frameworks, keeping the chat's state in sync with multiple outputs typically requires more manual wiring.
:::

> ## Custom tools

Reviewer: I'd be tempted to leave this for a future post, or just link to docs on the querychat site.

Author: I don't trust myself to come back and write a post on this, so I think I'll leave it 😅
